sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager.go

```diff
@@ -211,6 +211,16 @@ func (em *ElementManager) AddStage(ID string, inputIDs, outputIDs []string, side
 	for _, input := range inputIDs {
 		em.consumers[input] = append(em.consumers[input], ss.ID)
 	}
+
+	// In very rare cases, we can have a stage without any inputs, such as a flatten.
+	// In that case, there's nothing that will start the watermark refresh cycle,
+	// so we must do it here.
+	if len(inputIDs) == 0 {
```

Comment:
Non-PR blocking comment to validate my understanding. Do I understand correctly?

In the PTransform with no PCollection case, there are no inputIDs because the result of `internal.transformPreparer.PrepareTransform` returns an `internal.prepareResult` where `prepResult.SubbedComps.GetPcollections()` is empty.

Reply:
Effectively yes, but basically it's due to flatten being special: all Other Stages must have a single parallel input pCollection. Flattens may have any number, which apparently includes 0.

In this case, by construction, the flatten has no inputs to begin with.

Impulses and TestStream are also special: Impulses are required to kick off SDK side processing. TestStream is just weird.